_posts/fr/2018-02-11-throwable-error-exception.md

```diff
@@ -0,0 +1,374 @@
+          |- OutOfBoundsException extends RuntimeException
+          |- OverflowException extends RuntimeException
+          |- PDOException extends RuntimeException
+          |- RangeException extends RuntimeException
+          |- UnderflowException extends RuntimeException
+          |- UnexpectedValueException extends RuntimeException
+```
+
+> ⚠ Attention ! Vous ne pouvez implémenter `Throwable` qu'à travers `Error` et `Exception`. 
+> Sinon vous obtiendrez une erreur FATAL 
+> `PHP Fatal error:  Class MyClass cannot implement interface Throwable, extend Exception or Error instead`
+> Il est quand même possible d'étendre cette interface dans le domaine utilisateur.
+
+```php
+interface MyThrowable extends Throwable {
+    public function myCustomMethod();
+}
+
+class MyException extends Exception implements MyThrowable {
+    public function myCustomMethod()
+    {
+        // implement custom method code
+    }
+}
+```
+
+### Error
+
+`Error` est la classe de base de toutes les erreurs internes de PHP.
+
+Les plus communes sont :
+ - `ParseError`, qui est lancée quand on `require` ou `eval` un code qui contient une erreur de syntax.
+ - `TypeError`, qui est lancé quand le typehint d'un argument/retour d'une fonction n'est pas respectée. _Et également en `strict mode` quand on passe un nombre invalid d'arguments à une fonction native de PHP._
+
+_Vous pourriez être amenés à throw des `Error` dans votre code si par exemple vous parsez un fichier et qu'il contient une erreur de syntaxe.
+Ou si vous avez une fonction avec un nombre de paramètres variable et que le nombre/type d'arguments n'est pas correct._
+
+### Exception
+
+`Exception` est la classe de base de toutes les exceptions utilisateurs.
+
+Il est très fréquent de lancer ou créer des `Exception`. 
+C'est d'ailleurs sur le fonctionnement et l'utilisation des `Exception` que nous nous concentrer. 
+
+## Utilisation
+
+### Lancer une exception
+
+Pour lancer une exception, il suffit d'utiliser le mot clé `throw`.
+
+```php
+throw new Exception('Mon message d\'erreur.');
+echo "Affichage d'un contenu texte.";
+```
+
+> Il faut savoir qu'une `Exception` interrompt l'exécution des instructions suivantes.
+> Dans l'exemple le `echo` ne seras pas exécuté.
+
+### Attraper une exception
+
+Pour attraper et gérer l'exception, il faut utiliser la structure `try` `catch`.
+```php
+try {
+    if (!$_GET['titre']) {
+        throw new Exception('Impossible d'afficher le titre. Le titre est requis.);
+    }
+    echo $_GET['titre'];
+}
+catch (Exception $e) {
+    echo '⚠ Une exception est survenue : ' . $e->getMessage();
+}
+```
+> Dans cet exemple le script affichera le titre s'il est fourni
+> sinon il affichera le message d'erreur comme quoi il est obligatoire.
+
+Depuis PHP 7.1 il est également possible de spécifier plusieurs types d'`Exception` dans le catch en utilisant le caractère `|`
+
+```php
+try {
+    // Code
+}
+catch (OutOfBoundsException | LogicException $e) {
```

Comment:
Maybe you can talk about the priority ?

Explain (for example the importance of catching \Exception at the end ...) :

```
try {
   // Some code...
} catch (ExceptionType1 $e) {
   // Code to handle the exception
} catch (ExceptionType2 $e) {
   // Same code to handle the exception
} catch (Exception $e) {
   // ...
}
```

see https://wiki.php.net/rfc/multiple-catch

Reply:
yeah i forgot to talk about multiple catch but it's a different feature then `wildcard catch`
i'm gonna to add a paragraph to talk about multi catch thx :)